### PR TITLE
feat: LT-1586 - Licence duration view and permissions

### PR DESCRIPTION
**Pull request by a-bek**
1) Introduce new permission MANAGE_LICENCE_DURATION
2) Add duration field to BaseApplication
3) Create a finalise view, and disallow status to be set to finalised in ManageStatusView
4) Default durations logic (not in db)
5) Added is_eu flag to lite_content countries.csv as a datapoint to be stored in Country model

### CODE REVIEW

File: applications/libraries/licence.py [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+
+def get_default_duration(application):
+    """
+    Returns default duration in months
+
+    Rules defined in: https://uktrade.atlassian.net/browse/LT-1586
+    """
+
+    if application.export_type == ApplicationExportType.TEMPORARY:
+        return 1 * 12
```
Comment (BenAndrew-Kainos) [inline applications/libraries/licence.py]: Could you define these as constants at the top

File: applications/models.py [hunk 2]
```diff
@@ -23,6 +24,12 @@ class BaseApplication(Case):
     application_type = models.CharField(choices=ApplicationType.choices, default=None, max_length=50)
     activity = models.TextField(default=None, blank=True, null=True)
     usage = models.TextField(default=None, blank=True, null=True)
+    duration = models.IntegerField(
```
Comment (BenAndrew-Kainos) [inline applications/models.py]: Is duration applicable to an application or licence (should this not be the start of the licence model)?

File: applications/tests/tests_duration_view.py [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+        url = reverse("applications:duration", kwargs={"pk": self.open_application.pk})
+
+        country = CountryOnApplication.objects.get(application=self.open_application).country
+
+        # Non eu
+        country.is_eu = False
+        country.save()
+        response = self.client.get(url, **self.gov_headers)
+
+        self.assertEqual(response.json()["duration"], 5 * 12)
```
Comment (BenAndrew-Kainos) [inline applications/tests/tests_duration_view.py]: When you move the duration to constants, you can update these values to point to those constants (so if the default changes the tests wont break)

File: applications/tests/tests_duration_view.py [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+
+        # eu
+        country.is_eu = True
```
Comment (BenAndrew-Kainos) [inline applications/tests/tests_duration_view.py]: I know it's fussy but could you split the EU & non EU bits into separate tests and add to the get statement to get EU/Non-EU where applicable rather than overwriting

File: applications/tests/tests_finalise_application.py [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+        super().setUp()
+        self.standard_application = self.create_standard_application(self.organisation)
+        self.submit_application(self.standard_application)
+        self.url = reverse("applications:finalise", kwargs={"pk": self.standard_application.id})
+
+    def test_gov_super_user_finalise_application(self):
+        self.assertEqual(self.standard_application.duration, None)
+
+        # Give gov user super used role, to include reopen closed cases permission
+        self.gov_user.role = self.super_user_role
```
Comment (BenAndrew-Kainos) [inline applications/tests/tests_finalise_application.py]: Culd you be more specific and just add the 'include reopen closed cases' permission (proves that it is this permission that allows them)

File: applications/tests/tests_finalise_application.py [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+        self.gov_user.save()
+
+        data = {"duration": 12}
+        response = self.client.put(self.url, data=data, **self.gov_headers)
+
+        self.standard_application.refresh_from_db()
+        self.assertEqual(response.status_code, status.HTTP_200_OK)
+        self.assertEqual(self.standard_application.status, get_case_status_by_status(CaseStatusEnum.FINALISED))
+        self.assertEqual(self.standard_application.duration, data["duration"])
```
Comment (BenAndrew-Kainos) [inline applications/tests/tests_finalise_application.py]: Could you add a test for them not having the permission & assert 404

File: applications/views/applications.py [hunk 7]
```diff
@@ -241,25 +244,29 @@ class ApplicationManageStatus(APIView):
     def put(self, request, pk):
         application = get_application(pk)
 
-        data = request.data
-        new_status = data.get("status")
+        data = deepcopy(request.data)
```
Comment (BenAndrew-Kainos) [inline applications/views/applications.py]: Not sure deepcopy is required here (could you just use .copy() ?)

File: applications/views/applications.py [hunk 8]
```diff
@@ -241,25 +244,29 @@ class ApplicationManageStatus(APIView):
...
+
+        if data["status"] == CaseStatusEnum.FINALISED:
+            return JsonResponse(
+                data={"errors": ["Status cannot be set to finalised."]}, status=status.HTTP_400_BAD_REQUEST
```
Comment (BenAndrew-Kainos) [inline applications/views/applications.py]: Please move this error to content in the API section

File: applications/views/applications.py [hunk 9]
```diff
@@ -241,25 +244,29 @@ class ApplicationManageStatus(APIView):
...
+            )
 
         if isinstance(request.user, ExporterUser):
             if request.user.organisation.id != application.organisation.id:
                 raise PermissionDenied()
 
-            if not can_status_can_be_set_by_exporter_user(application.status.status, new_status):
+            if not can_status_can_be_set_by_exporter_user(application.status.status, data.get("status")):
                 return JsonResponse(
                     data={"errors": ["Status cannot be set by Exporter user."]}, status=status.HTTP_400_BAD_REQUEST
```
Comment (BenAndrew-Kainos) [inline applications/views/applications.py]: Same as above

File: applications/views/applications.py [hunk 10]
```diff
@@ -241,25 +244,29 @@ class ApplicationManageStatus(APIView):
...
                 )
         else:
-            if not can_status_can_be_set_by_gov_user(request.user, application.status.status, new_status):
+            if not can_status_can_be_set_by_gov_user(request.user, application.status.status, data.get("status")):
                 return JsonResponse(
                     data={"errors": ["Status cannot be set by Gov user."]}, status=status.HTTP_400_BAD_REQUEST
```
Comment (BenAndrew-Kainos) [inline applications/views/applications.py]: Same as above

File: applications/views/applications.py [hunk 11]
```diff
@@ -241,25 +244,29 @@ class ApplicationManageStatus(APIView):
...
                 )
 
-        new_status = get_case_status_by_status(new_status)
-        request.data["status"] = str(new_status.pk)
+        case_status = get_case_status_by_status(data["status"])
```
Comment (BenAndrew-Kainos) [inline applications/views/applications.py]: You have done a mixture of data["status"] & data.get("status"). Could you make this consistent

File: applications/views/applications.py [hunk 12]
```diff
@@ -273,7 +280,65 @@ def put(self, request, pk):
...
+    @transaction.atomic
+    def put(self, request, pk):
+        """
+        Finalise an application
+        """
+        application = get_application(pk)
+
+        if not can_status_can_be_set_by_gov_user(request.user, application.status.status, CaseStatusEnum.FINALISED):
+            return JsonResponse(
+                data={"errors": ["Status cannot be set by Gov user."]}, status=status.HTTP_400_BAD_REQUEST
```
Comment (BenAndrew-Kainos) [inline applications/views/applications.py]: Move string to content

File: cases/tests/tests_final_decisions.py [hunk 13]
```diff
@@ -33,10 +33,10 @@ def test_can_record_final_decision_with_correct_permissions(self):
         self.gov_user.role = role
         self.gov_user.save()
 
-        data = {"status": CaseStatusEnum.FINALISED}
+        data = {"duration": 10}
```
Comment (BenAndrew-Kainos) [inline cases/tests/tests_final_decisions.py]: Have you not just changed the purpose of this test (from checking FINALISED to a duration)

**Comment by BenAndrew-Kainos** [comment]
I'm a bit concerned about the work done to 'finalised' statuses in this story. I don't think it's part of the scope

File: applications/models.py (hunk 2 above)
Comment (a-bek) [reply applications/models.py]: Duration is part of the application, like other things, that end up in the licence. I considered starting a Licence model, but I think its best to wait for a story that requires a Licence model.

File: applications/views/applications.py (hunk 7 above)
Comment (a-bek) [reply applications/views/applications.py]: copy won't truly copy mutable objects. With copy: `id(data) != id(request.data)` but `id(data['mutable_list']) == id(request.data['mutable_list'])`

File: applications/models.py (hunk 2 above)
Comment (stephanieboyd) [reply applications/models.py]: During the tech discussion for this ticket, it was agreed that "duration" should be on a licence. It was also agreed (and looking at the lack of subtasks) that we wouldn't create the licence model or store the value anywhere just yet. Please remove from application, this is not the correct place for this property :)

File: cases/tests/tests_final_decisions.py (hunk 13 above)
Comment (a-bek) [reply cases/tests/tests_final_decisions.py]: No, previously finalising a case was using the `manage_status` view. `manage_status` view can no longer finalise a case, and finalising has a view of its own, which this is now testing and only expects a `duration` param

**Comment by a-bek** [comment]
> I'm a bit concerned about the work done to 'finalised' statuses in this story. I don't think it's part of the scope

@BenAndrew-Kainos Could you elaborate?

File: applications/models.py [hunk 14]
```diff
@@ -23,6 +19,7 @@ class BaseApplication(Case):
     application_type = models.CharField(choices=ApplicationType.choices, default=None, max_length=50)
     activity = models.TextField(default=None, blank=True, null=True)
     usage = models.TextField(default=None, blank=True, null=True)
+    duration = models.IntegerField(default=None, null=True, help_text="Set when application finalised")
```
Comment (stephanieboyd) [inline applications/models.py]: Please rename to licence_duration

File: applications/views/applications.py [hunk 15]
```diff
@@ -273,7 +283,72 @@ def put(self, request, pk):
...
+                actor=request.user,
+                verb=AuditType.UPDATED_STATUS,
+                target=application.get_case(),
+                payload={"status": CaseStatusEnum.human_readable(CaseStatusEnum.FINALISED)},
+            )
+
+        return JsonResponse(data={}, status=status.HTTP_200_OK)
+
+
+class DurationView(APIView):
```
Comment (stephanieboyd) [inline applications/views/applications.py]: Perhaps rename to:  ApplicationDurationView

File: applications/views/applications.py [hunk 16]
```diff
@@ -273,7 +283,72 @@ def put(self, request, pk):
...
             verb=AuditType.UPDATED_STATUS,
             target=application.get_case(),
-            payload={"status": CaseStatusEnum.human_readable(new_status.status)},
+            payload={"status": CaseStatusEnum.human_readable(case_status.status)},
         )
 
         return JsonResponse(data={}, status=status.HTTP_200_OK)
+
+
+class FinaliseView(APIView):
```
Comment (stephanieboyd) [inline applications/views/applications.py]: Perhaps rename to: ApplicationFinaliseView

File: conf/constants.py [hunk 17]
```diff
@@ -11,6 +11,7 @@ class GovPermissions(Enum):
     CONFIRM_OWN_ADVICE = "Confirm own advice"
     CONFIGURE_TEMPLATES = "Create and edit templates"
     REOPEN_CLOSED_CASES = "Can re-open closed cases"
+    MANAGE_LICENCE_DURATION = "Can edit licence duration"
```
Comment (stephanieboyd) [inline conf/constants.py]: Should this permission not reflect overriding like mentioned in the ticket. 

`Override_default_licence_duration`

File: applications/models.py (hunk 2 above)
Comment (BenAndrew-Kainos) [reply applications/models.py]: As per our call we are leaving this. However, please rename to `licence_duration` and rebuild migrations / update accordingly

File: applications/enums.py [hunk 18]
```diff
@@ -28,3 +28,8 @@ class ApplicationExportLicenceOfficialType:
         (YES, "Yes"),
         (NO, "No"),
     ]
+
+
+class Duration:
```
Comment (BenAndrew-Kainos) [inline applications/enums.py]: Please rename to LicenceDuration

File: applications/models.py (hunk 14 above)
Comment (BenAndrew-Kainos) [reply applications/models.py]: ^^

File: applications/tests/tests_duration_view.py (hunk 4 above)
Comment (BenAndrew-Kainos) [reply applications/tests/tests_duration_view.py]: Could you add `is_eu=True` to the get instead of overwriting

File: applications/tests/tests_finalise_application.py [hunk 19]
```diff
@@ -0,0 +1,72 @@
...
+    def test_gov_super_invalid_role(self):
+        self.gov_user.role = self.role
+        self.gov_user.role.permissions.set([])
+        self.gov_user.save()
+
+        data = {"duration": 13}
+        response = self.client.put(self.url, data=data, **self.gov_headers)
+
+        self.assertEqual(response.status_code, status.HTTP_403_FORBIDDEN)
+        self.assertEqual(response.json(), {"errors": {"error": "You don't have permission to do that action"}})
```
Comment (BenAndrew-Kainos) [inline applications/tests/tests_finalise_application.py]: Please point the error to the string in lite_content

File: applications/tests/tests_finalise_application.py [hunk 20]
```diff
@@ -0,0 +1,72 @@
...
+        self.gov_user.role.permissions.set(
+            [GovPermissions.MANAGE_FINAL_ADVICE.name, GovPermissions.MANAGE_LICENCE_DURATION.name]
+        )
+        self.gov_user.save()
+
+        data = {"duration": Duration.MAX + 1}
+        response = self.client.put(self.url, data=data, **self.gov_headers)
+
+        self.assertEqual(response.status_code, status.HTTP_400_BAD_REQUEST)
+        self.assertEqual(response.json(), {"errors": {"non_field_errors": ["Duration 1000 not in range [1-999]"]}})
```
Comment (BenAndrew-Kainos) [inline applications/tests/tests_finalise_application.py]: Please point the error to the string in lite_content

File: applications/tests/tests_finalise_application.py [hunk 21]
```diff
@@ -0,0 +1,72 @@
...
+
+    def test_default_duration_can_be_set_without_role(self):
+        self.gov_user.role = self.role
+        self.gov_user.role.permissions.set([GovPermissions.MANAGE_FINAL_ADVICE.name])
+        self.gov_user.save()
+
+        data = {"duration": get_default_duration(self.standard_application)}
+        response = self.client.put(self.url, data=data, **self.gov_headers)
+        self.standard_application.refresh_from_db()
```
Comment (BenAndrew-Kainos) [inline applications/tests/tests_finalise_application.py]: This test wouldn't catch if the duration wasn't updated. It's only asserting that the duration sent back matches what is in the object, not what was sent in the put request. 
Please change to `self.assertEqual(get_default_duration(self.standard_application), data["duration"])`

File: applications/views/applications.py (hunk 7 above)
Comment (BenAndrew-Kainos) [reply applications/views/applications.py]: ok

File: cases/tests/tests_final_decisions.py (hunk 13 above)
Comment (BenAndrew-Kainos) [reply cases/tests/tests_final_decisions.py]: Could you write a docstring & rename the unit test to make this clear

File: applications/tests/tests_finalise_application.py (hunk 21 above)
Comment (a-bek) [reply applications/tests/tests_finalise_application.py]: The test asserts that a user can finalise the application using the default duration, without having manage_licence_duration permission.

Also, data["duration"] is `get_default_duration(self.standard_application)`

File: applications/views/applications.py (hunk 16 above)
Comment (a-bek) [reply applications/views/applications.py]: It's already within the `application` namespace. It's better practice to avoid name smurfing.

File: conf/constants.py (hunk 17 above)
Comment (a-bek) [reply conf/constants.py]: I think MANAGE_LICENCE_DURATION is more suitable, what if the user wants to edit a duration that the user set when reissuing a licence, ie, editing a non-default duration. Then `Override_default_licence_duration` wouldn't be a clear name.

File: applications/tests/tests_duration_view.py (hunk 4 above)
Comment (a-bek) [reply applications/tests/tests_duration_view.py]: Just explicitly setting it for testing purposes.

File: applications/views/applications.py (hunk 16 above)
Comment (BenAndrew-Kainos) [reply applications/views/applications.py]: I'm still against us having this extra view for when cases are finalised. Is there a reason we cannot add this logic to the existing update status view?

File: applications/tests/tests_finalise_application.py (hunk 19 above)
Comment (BenAndrew-Kainos) [reply applications/tests/tests_finalise_application.py]: Do not resolve a point without reason or action please

File: applications/tests/tests_finalise_application.py (hunk 20 above)
Comment (BenAndrew-Kainos) [reply applications/tests/tests_finalise_application.py]: Same here

**Comment by BenAndrew-Kainos** [comment]
In future please leave a comment why it's resolved (i.e. done)

**Comment by a-bek** [comment]
@BenAndrew-Kainos 
`But this means we don't test that we have any EU/ non-EU countries`
We're not testing data, we're testing the behaviour of the duration view for a country in EU.

`I'm still against us having this extra view for when cases are finalised. Is there a reason we cannot add this logic to the existing update status view?`
It's a seperation of concerns, manage status view is simply to edit a status. Finalising an application is a process in itself, I really don't think that belongs as and edge case in the manage_status view.

**Comment by jfaracik** [comment]
> It's a seperation of concerns, manage status view is simply to edit a status. Finalising an application is a process in itself, I really don't think that belongs as and edge case in the manage_status view.

Agreed with this.

**Comment by BenAndrew-Kainos** [comment]
> @BenAndrew-Kainos
> `But this means we don't test that we have any EU/ non-EU countries`
> We're not testing data, we're testing the behaviour of the duration view for a country in EU.

I know it's not the direct purpose of this test but it is important to verify EU countries are being loaded somewhere. If you don't want to do it here for some reason, update the seeding unit test to do so

**Comment by BenAndrew-Kainos** [comment]
I'm still against us having this extra view for when cases are finalised. Is there a reason we cannot add this logic to the existing update status view?
It's a seperation of concerns, manage status view is simply to edit a status. Finalising an application is a process in itself, I really don't think that belongs as and edge case in the manage_status view.

That's a fair point but to me it's more important that you have a consistent API design. I would expect as a dev that all updates to status go via the updateStatus PUT request. Maintaining two endpoints for updating status seems unnecessary to me. 

@stephanieboyd you've done the most work with statuses before. What do you reckon?

**Comment by a-bek** [comment]
```That's a fair point but to me it's more important that you have a consistent API design. I would expect as a dev that all updates to status go via the updateStatus PUT request. Maintaining two endpoints for updating status seems unnecessary to me.```

@BenAndrew-Kainos having `/finalise/` is actually consistent with API design. It is like `/submit/`as it edits the status to `SUBMITTED` and any other logic to do with submitting. Now if we went down the route of baking everything in manage_status, we would have to move the existing `/submit/` logic to manage_status as well, and eventually you end up with a giant manage_status view that handles edge cases.

Submit and Finalise are two actions that, yes, require managing status, but also require submit and finalise specific logic, so they warrant their own view.

File: applications/libraries/licence.py [hunk 22]
```diff
@@ -0,0 +1,33 @@
...
+    """
+    Returns default duration in months
+
+    Rules defined in: https://uktrade.atlassian.net/browse/LT-1586
+    """
+
+    if application.export_type == ApplicationExportType.TEMPORARY:
+        return DefaultDuration.TEMPORARY
+
+    if (
```
Comment (Mark-Spiers-kainos) [inline applications/libraries/licence.py]: can you make these elifs?

File: applications/views/applications.py (hunk 16 above)
Comment (stephanieboyd) [reply applications/views/applications.py]: You're inconsistent with how the team name things, and within this file already there is a pattern. Please rename

File: applications/urls.py [hunk 23]
```diff
@@ -19,6 +19,8 @@
     path("", applications.ApplicationList.as_view(), name="applications"),
     path("<uuid:pk>/", applications.ApplicationDetail.as_view(), name="application"),
     path("<uuid:pk>/submit/", applications.ApplicationSubmission.as_view(), name="application_submit",),
+    path("<uuid:pk>/finalise/", applications.FinaliseView.as_view(), name="finalise"),
```
Comment (Mark-Spiers-kainos) [inline applications/urls.py]: can you change name to "finalise_application" or something similar? same with duration

File: applications/libraries/licence.py (hunk 22 above)
Comment (a-bek) [reply applications/libraries/licence.py]: done, but it doesn't make a difference.

File: applications/urls.py (hunk 23 above)
Comment (a-bek) [reply applications/urls.py]: I prefer 
`/applications/1/finalise/`
to
`/applications/1/finalise_application/`

https://devcards.io/smurf-naming-convention

File: applications/tests/tests_duration_view.py (hunk 4 above)
Comment (a-bek) [reply applications/tests/tests_duration_view.py]: Added test to static

File: applications/serializers/generic_application.py [hunk 24]
```diff
@@ -187,10 +191,23 @@ def update(self, instance, validated_data):
...
         return instance
+
+    def validate(self, data):
+        """
+        Check that the start is before the stop.
+        """
+        if data.get("licence_duration") is not None and (
+            data["licence_duration"] > LicenceDuration.MAX.value or data["licence_duration"] < LicenceDuration.MIN.value
+        ):
+            _min = LicenceDuration.MIN.value
```
Comment (BenAndrew-Kainos) [inline applications/serializers/generic_application.py]: Where are these used??

File: applications/views/applications.py [hunk 25]
```diff
@@ -273,7 +283,72 @@ def put(self, request, pk):
...
+        data = deepcopy(request.data)
+
+        default_licence_duration = get_default_duration(application)
+
+        if (
+                data.get("licence_duration") is not None and
+                str(data["licence_duration"]) != str(default_licence_duration) and
+                not request.user.has_permission(GovPermissions.MANAGE_LICENCE_DURATION)
+        ):
+            #
```
Comment (BenAndrew-Kainos) [inline applications/views/applications.py]: Remove

File: applications/views/applications.py [hunk 26]
```diff
@@ -273,7 +283,72 @@ def put(self, request, pk):
...
+        serializer.save()
+
+        audit_trail_service.create(
+            actor=request.user,
+            verb=AuditType.FINALISED_APPLICATION,
+            target=application.get_case(),
+            payload={"licence_duration": serializer.validated_data["licence_duration"]},
+        )
+
+        return JsonResponse(data={}, status=status.HTTP_200_OK)
```
Comment (BenAndrew-Kainos) [inline applications/views/applications.py]: This is a PUT endpoint so should return the updated object

File: applications/views/applications.py (hunk 15 above)
Comment (BenAndrew-Kainos) [reply applications/views/applications.py]: Resolved without comment or action?

File: static/countries/enums.py [hunk 27]
```diff
@@ -0,0 +1,5 @@
+from enum import Enum
+
+
+class Misc(Enum):
```
Comment (BenAndrew-Kainos) [inline static/countries/enums.py]: Don't like this at all. This enum will be left behind over time with countries being updated. Please remove

File: static/countries/models.py [hunk 28]
```diff
@@ -1,10 +1,16 @@
...
 class Country(models.Model):
     id = models.CharField(primary_key=True, editable=False, max_length=10)  # Country Code
     name = models.CharField(max_length=100)
     type = models.CharField(max_length=100)
     flags = models.ManyToManyField(Flag, related_name="countries")
+    is_eu = models.BooleanField()
+
+    @classmethod
+    def eu_count(cls):
+        return Misc.EU_COUNT.value
```
Comment (BenAndrew-Kainos) [inline static/countries/models.py]: Why do you need this count. (Also change to `Country.objects.filter(is_eu=True).count()`)

File: static/management/commands/seedcountries.py [hunk 29]
```diff
@@ -26,4 +27,5 @@ def operation(self, *args, **options):
 class SeedCountriesTests(SeedCommandTest):
     def test_seed_countries(self):
         self.seed_command(Command)
-        self.assertTrue(Country.objects.count() == len(Command.read_csv(COUNTRIES_FILE)))
+        self.assertEqual(Country.objects.count(), len(Command.read_csv(COUNTRIES_FILE)))
+        self.assertEqual(Country.objects.filter(is_eu=True).count(), Misc.EU_COUNT.value)
```
Comment (BenAndrew-Kainos) [inline static/management/commands/seedcountries.py]: Instead of this compare the number of occurrences in the csv to the number of objects. Also only read the csv once (every time you call `Command.read_csv(COUNTRIES_FILE))` you're unnecessarily re-reading the csv)

File: cases/tests/tests_final_decisions.py (hunk 13 above)
Comment (BenAndrew-Kainos) [reply cases/tests/tests_final_decisions.py]: Resolved without comment or action (again)

File: applications/views/applications.py (hunk 15 above)
Comment (a-bek) [reply applications/views/applications.py]: It's already within the application namespace. It's better practice to avoid name smurfing.

File: static/countries/enums.py (hunk 27 above)
Comment (a-bek) [reply static/countries/enums.py]: Without over-engineering, this is a simple way to validate the EU countries data like you asked. There is nothing wrong with using this as a miscellaneous value, if the actual number of EU countries are updated in the csv this will be caught.

File: static/countries/models.py (hunk 28 above)
Comment (a-bek) [reply static/countries/models.py]: Removed

File: static/management/commands/seedcountries.py (hunk 29 above)
Comment (a-bek) [reply static/management/commands/seedcountries.py]: Reading the CSV is only being called once in the test function block, and is used to validate `seed_command` does what it is supposed to do

File: applications/views/applications.py (hunk 16 above)
Comment (a-bek) [reply applications/views/applications.py]: Yes I know there some kind of a pattern already, but even the existing pattern is inconsistent. Generally its recommended to avoid name smurfing, so that SHOULD be the practise going forward, but sure I'll add the Application prefix.

File: static/management/commands/seedcountries.py (hunk 29 above)
Comment (BenAndrew-Kainos) [reply static/management/commands/seedcountries.py]: It was being used twice when I reviewed but looks good now

File: static/countries/enums.py (hunk 27 above)
Comment (BenAndrew-Kainos) [reply static/countries/enums.py]: The problem is content have the responsibility to keep CSV's up to date but not enums. If a country becomes EU or leaves then the tests will fail because this number is incorrect. I would replace all instances of this with a .count() on countries.

**Comment by BenAndrew-Kainos** [comment]
Approving subject to the Misc Enum being deleted!!!

File: static/countries/enums.py (hunk 27 above)
Comment (a-bek) [reply static/countries/enums.py]: Thats expected behaviour of the test? It validates content have added the right number of EU countries and it validates that EU countries exist? I thought this is what you wanted when you said:

`But this means we don't test that we have any EU/ non-EU countries`

Updated and removed again.
